pkg/releaseindex/release_index.go

```diff
@@ -0,0 +1,167 @@
+	// installations repository.
+	GithubToken string
+	Logger      micrologger.Logger
+}
+
+// TODO consider this to be extracted as a library.
+//
+//	See https://github.com/giantswarm/giantswarm/issues/4656
+//
+type ReleaseIndex struct {
+	githubToken string
+	logger      micrologger.Logger
+
+	cache    map[Provider][]versionbundle.IndexRelease
+	cacheMux *sync.RWMutex
+}
+
+func New(config Config) (*ReleaseIndex, error) {
+	if config.GithubToken == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.GithubToken must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	r := &ReleaseIndex{
+		githubToken: config.GithubToken,
+		logger:      config.Logger,
+
+		cache:    map[Provider][]versionbundle.IndexRelease{},
+		cacheMux: &sync.RWMutex{},
+	}
+
+	return r, nil
+}
+
+// GetVersion returns a component version for a given type and provider.
+//
+// NOTE: Release index is cached for a provider after first call. This makes it
+// cheap to call this method multiple times with the same provider. But changes
+// to the installations repository will not be visible.
+func (r *ReleaseIndex) GetVersion(ctx context.Context, provider Provider, versionType VersionType, component Authority) (string, error) {
+	index, err := r.getIndex(ctx, provider)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	version, err := getVersion(index, versionType, component)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	return version, nil
+}
+
+func (r *ReleaseIndex) getFileContent(ctx context.Context, owner, repo, path string) (string, error) {
+	var client *github.Client
+	{
+		t := &oauth2.Token{
+			AccessToken: r.githubToken,
+		}
+
+		sts := oauth2.StaticTokenSource(t)
+		tc := oauth2.NewClient(ctx, sts)
+
+		client = github.NewClient(tc)
+	}
+
+	opt := &github.RepositoryContentGetOptions{}
+	repoContent, _, _, err := client.Repositories.GetContents(ctx, owner, repo, path, opt)
+
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	content, err := repoContent.GetContent()
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	return content, nil
+}
+
+func (r *ReleaseIndex) getIndex(ctx context.Context, provider Provider) ([]versionbundle.IndexRelease, error) {
+	var err error
+
+	{
+		r.cacheMux.RLock()
+		index, ok := r.cache[provider]
+		r.cacheMux.RUnlock()
+
+		if ok {
+			return index, nil
+		}
+	}
+
+	var indexYAML string
+	{
+		owner := "giantswarm"
+		repo := "installations"
+		path := fmt.Sprintf("release/provider/%s.yaml", provider.String())
+
+		indexYAML, err = r.getFileContent(ctx, owner, repo, path)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var index []versionbundle.IndexRelease
+	{
+		err := yaml.Unmarshal([]byte(indexYAML), &index)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	{
+		r.cacheMux.Lock()
+		r.cache[provider] = index
+		r.cacheMux.Unlock()
```

Comment:
I do not get the point about using a lock here if it is released in between read and write operations.
Then this code is exposed to race conditions where `r.cache[provider]` could get overwritten more than once.

Reply:
It isn't released in between:

1. Lock.
2. Write the data.
3. Unlock.

Do I miss something?

Reply:
Oh I know what you mean. Yes there is such possibility. So it's probably better to to just use a lock with defer at the beggining of the function.

Reply:
I've fixed that. That was a good catch. Maybe that's also why @tuommaki was confused.